File: mobile/modules/frame-preview/android/src/main/java/com/mentra/framepreview/PreviewTrace.kt
package com.mentra.framepreview

import android.os.SystemClock

/**
 * Native side of the `PREVIEW_TRACE` lifecycle log.
 *
 * Every line is `[PREVIEW_TRACE] previewTraceId=… phase=… t=… key=value…`, the same shape the
 * host and SDK use, so one grep for a trace id reconstructs a preview across layers. Lines go to
 * logcat and to [sink], which the module forwards to JS so they land in bug-report logs.
 *
 * Lifecycle only: nothing here is called per frame. Repeatable warnings go through [warnLimited],
 * which logs the first occurrence and then one count per 10 s window.
 */
class PreviewTrace(
  private val sink: (level: Level, line: String) -> Unit,
  private val clockMs: () -> Long = { SystemClock.elapsedRealtime() },
) {
  enum class Level(val wire: String) { INFO("info"), WARN("warn") }

  @Volatile var traceId: String = ""

  /** Correlation ids stamped on every line when set. */
  @Volatile var docGen: Int? = null

  @Volatile var tapGeneration: Long? = null

  private val limiter = PreviewRateLimiter()

  fun info(phase: String, fields: Map<String, Any?> = emptyMap()) = emit(Level.INFO, phase, fields)

  fun warn(phase: String, fields: Map<String, Any?> = emptyMap()) = emit(Level.WARN, phase, fields)

  /** Warn once, then at most once per window with the number of repeats suppressed. */
  fun warnLimited(key: String, phase: String, fields: Map<String, Any?> = emptyMap()) {
    val suppressed = limiter.check(key, clockMs()) ?: return
    emit(Level.WARN, phase, if (suppressed > 0) fields + ("suppressed" to suppressed) else fields)
  }

  /** Emit pending suppressed counts whose window has closed. Called from the 1 Hz tick. */
  fun flushLimited() {
    for ((key, count) in limiter.flush(clockMs())) {
      emit(Level.WARN, "repeated_warning", mapOf("key" to key, "suppressed" to count))
    }
  }

  private fun emit(level: Level, phase: String, fields: Map<String, Any?>) {
    val ids = linkedMapOf<String, Any?>()
    docGen?.let { ids["docGen"] = it }
    tapGeneration?.let { ids["gen"] = it }
    val line = format(traceId, phase, clockMs(), ids + fields)
    try {
      sink(level, line)
    } catch (ignored: Throwable) {
      // A logging failure must not change what the preview does.
    }
  }

  companion object {
    /** Grep marker. Never build this by concatenation — a single grep must be exhaustive. */
    const val MARKER = "PREVIEW_TRACE"

    private const val REDACTED = "<redacted>"

    /** Keys whose values never reach the log, matched case-insensitively as substrings. */
    private val SENSITIVE_KEYS = listOf("token", "secret", "password", "credential", "authorization", "meetingurl")

    fun format(traceId: String, phase: String, tMs: Long, fields: Map<String, Any?>): String {
      val builder = StringBuilder("[").append(MARKER).append("]")
      if (traceId.isNotEmpty()) builder.append(" previewTraceId=").append(traceId)
      builder.append(" phase=").append(phase).append(" t=").append(tMs)
      for ((key, value) in fields) {
        if (value == null) continue
        builder.append(' ').append(key).append('=').append(render(sanitize(key, value)))
      }
      return builder.toString()
    }

    fun sanitize(key: String, value: Any): Any {
      val lower = key.lowercase()
      if (SENSITIVE_KEYS.any { lower.contains(it) }) return REDACTED
      return if (value is String) stripUrlSecrets(value) else value
    }

    /** Drop `?query`, `#fragment` and `user:pass@` from anything URL-shaped. */
    fun stripUrlSecrets(text: String): String {
      val scheme = text.indexOf("://")
      if (scheme < 0) return text
      var result = text
      val query = result.indexOf('?')
      if (query >= 0) result = result.substring(0, query) + "?" + REDACTED
      val fragment = result.indexOf('#')
      if (fragment >= 0) result = result.substring(0, fragment)
      val at = result.indexOf('@', scheme + 3)
      if (at >= 0) result = result.substring(0, scheme + 3) + REDACTED + "@" + result.substring(at + 1)
      return result
    }

    private fun render(value: Any): String {
      val text = when (value) {
        is Map<*, *> -> value.entries.joinToString(",", "{", "}") { "${it.key}:${it.value}" }
        else -> value.toString()
      }
      return if (text.any { it.isWhitespace() || it == '"' }) "\"" + text.replace("\"", "'") + "\"" else text
    }
  }
}

/** First occurrence passes, repeats inside a 10 s window are counted and reported once. */
class PreviewRateLimiter(private val windowMs: Long = 10_000) {
  private class Entry(var windowStartMs: Long, var suppressed: Int)

  private val entries = HashMap<String, Entry>()

  /** Suppressed count to report with this occurrence (0 for the first), or null to stay quiet. */
  @Synchronized
  fun check(key: String, nowMs: Long): Int? {
    val entry = entries[key]
    if (entry == null) {
      entries[key] = Entry(nowMs, 0)
      return 0
    }
    if (nowMs - entry.windowStartMs < windowMs) {
      entry.suppressed += 1
      return null
    }
    val suppressed = entry.suppressed
    entry.windowStartMs = nowMs
    entry.suppressed = 0
    return suppressed
  }

  /** Counts for windows that closed with repeats nobody has reported yet. */
  @Synchronized
  fun flush(nowMs: Long): List<Pair<String, Int>> {
    val due = mutableListOf<Pair<String, Int>>()
    for ((key, entry) in entries) {
      if (entry.suppressed > 0 && nowMs - entry.windowStartMs >= windowMs) {
        due += key to entry.suppressed
        entry.windowStartMs = nowMs
        entry.suppressed = 0
      }
    }
    return due
  }
}

File: mobile/modules/frame-preview/ios/PreviewKit/Sources/FramePreviewCore/PreviewTrace.swift
import Foundation

/// Native side of the `PREVIEW_TRACE` lifecycle log.
///
/// Every line is `[PREVIEW_TRACE] previewTraceId=… phase=… t=… key=value…`, the same shape the
/// host and SDK use, so one grep for a trace id reconstructs a preview across layers. Lines go to
/// the sink, which the module sends to NSLog and to JS so they land in bug-report logs.
///
/// Lifecycle only: nothing here is called per frame. Repeatable warnings go through
/// `warnLimited`, which logs the first occurrence and then one count per 10 s window.
public final class PreviewTrace {
  public enum Level: String, Sendable {
    case info
    case warn
  }

  /// Grep marker. Never build this by concatenation — a single grep must be exhaustive.
  public static let marker = "PREVIEW_TRACE"
  private static let redacted = "<redacted>"
  private static let sensitiveKeys = ["token", "secret", "password", "credential", "authorization", "meetingurl"]

  private let sink: (Level, String) -> Void
  private let clockMs: () -> Int64
  private let lock = NSLock()
  private let limiter = PreviewRateLimiter()
  private var ids: (traceId: String, docGen: Int?, tapGeneration: UInt64?) = ("", nil, nil)

  public init(
    sink: @escaping (Level, String) -> Void,
    clockMs: @escaping () -> Int64 = { Int64(DispatchTime.now().uptimeNanoseconds / 1_000_000) }
  ) {
    self.sink = sink
    self.clockMs = clockMs
  }

  public var traceId: String {
    get { lock.lock(); defer { lock.unlock() }; return ids.traceId }
    set { lock.lock(); ids.traceId = newValue; lock.unlock() }
  }

  public var docGen: Int? {
    get { lock.lock(); defer { lock.unlock() }; return ids.docGen }
    set { lock.lock(); ids.docGen = newValue; lock.unlock() }
  }

  public var tapGeneration: UInt64? {
    get { lock.lock(); defer { lock.unlock() }; return ids.tapGeneration }
    set { lock.lock(); ids.tapGeneration = newValue; lock.unlock() }
  }

  public func info(_ phase: String, _ fields: KeyValuePairs<String, Any?> = [:]) {
    emit(.info, phase, fields.map { ($0.key, $0.value) })
  }

  public func warn(_ phase: String, _ fields: KeyValuePairs<String, Any?> = [:]) {
    emit(.warn, phase, fields.map { ($0.key, $0.value) })
  }

  /// Warn once, then at most once per window with the number of repeats suppressed.
  public func warnLimited(_ key: String, _ phase: String, _ fields: KeyValuePairs<String, Any?> = [:]) {
    guard let suppressed = limiter.check(key, nowMs: clockMs()) else { return }
    var pairs = fields.map { ($0.key, $0.value) }
    if suppressed > 0 { pairs.append(("suppressed", suppressed)) }
    emit(.warn, phase, pairs)
  }

  /// Emit pending suppressed counts whose window has closed. Called from the 1 Hz tick.
  public func flushLimited() {
    for (key, count) in limiter.flush(nowMs: clockMs()) {
      emit(.warn, "repeated_warning", [("key", key), ("suppressed", count)])
    }
  }

  private func emit(_ level: Level, _ phase: String, _ fields: [(String, Any?)]) {
    lock.lock()
    let current = ids
    lock.unlock()
    var all: [(String, Any?)] = []
    if let docGen = current.docGen { all.append(("docGen", docGen)) }
    if let generation = current.tapGeneration { all.append(("gen", generation)) }
    all.append(contentsOf: fields)
    sink(level, Self.format(traceId: current.traceId, phase: phase, tMs: clockMs(), fields: all))
  }

  public static func format(traceId: String, phase: String, tMs: Int64, fields: [(String, Any?)]) -> String {
    var line = "[\(marker)]"
    if !traceId.isEmpty { line += " previewTraceId=\(traceId)" }
    line += " phase=\(phase) t=\(tMs)"
    for (key, value) in fields {
      guard let value else { continue }
      line += " \(key)=\(render(sanitize(key: key, value: value)))"
    }
    return line
  }

  public static func sanitize(key: String, value: Any) -> Any {
    let lower = key.lowercased()
    if sensitiveKeys.contains(where: { lower.contains($0) }) { return redacted }
    if let text = value as? String { return stripUrlSecrets(text) }
    return value
  }

  /// Drop `?query`, `#fragment` and `user:pass@` from anything URL-shaped.
  public static func stripUrlSecrets(_ text: String) -> String {
    guard let scheme = text.range(of: "://") else { return text }
    var result = text
    if let query = result.firstIndex(of: "?") { result = String(result[..<query]) + "?" + redacted }
    if let fragment = result.firstIndex(of: "#") { result = String(result[..<fragment]) }
    let hostStart = result.index(scheme.lowerBound, offsetBy: 3)
    if let at = result[hostStart...].firstIndex(of: "@") {
      result = String(result[..<hostStart]) + redacted + "@" + String(result[result.index(after: at)...])
    }
    return result
  }

  private static func render(_ value: Any) -> String {
    let text: String
    if let map = value as? [String: Any] {
      text = "{" + map.keys.sorted().map { "\($0):\(map[$0].map { "\($0)" } ?? "")" }.joined(separator: ",") + "}"
    } else {
      text = "\(value)"
    }
    return text.contains(where: { $0.isWhitespace || $0 == "\"" })
      ? "\"" + text.replacingOccurrences(of: "\"", with: "'") + "\""
      : text
  }
}

/// First occurrence passes, repeats inside a 10 s window are counted and reported once.
public final class PreviewRateLimiter {
  private struct Entry {
    var windowStartMs: Int64
    var suppressed: Int
  }

  private let windowMs: Int64
  private let lock = NSLock()
  private var entries: [String: Entry] = [:]

  public init(windowMs: Int64 = 10000) {
    self.windowMs = windowMs
  }

  /// Suppressed count to report with this occurrence (0 for the first), or nil to stay quiet.
  public func check(_ key: String, nowMs: Int64) -> Int? {
    lock.lock()
    defer { lock.unlock() }
    guard var entry = entries[key] else {
      entries[key] = Entry(windowStartMs: nowMs, suppressed: 0)
      return 0
    }
    if nowMs - entry.windowStartMs < windowMs {
      entry.suppressed += 1
      entries[key] = entry
      return nil
    }
    let suppressed = entry.suppressed
    entries[key] = Entry(windowStartMs: nowMs, suppressed: 0)
    return suppressed
  }

  /// Counts for windows that closed with repeats nobody has reported yet.
  public func flush(nowMs: Int64) -> [(String, Int)] {
    lock.lock()
    defer { lock.unlock() }
    var due: [(String, Int)] = []
    for (key, entry) in entries where entry.suppressed > 0 && nowMs - entry.windowStartMs >= windowMs {
      due.append((key, entry.suppressed))
      entries[key] = Entry(windowStartMs: nowMs, suppressed: 0)
    }
    return due.sorted { $0.0 < $1.0 }
  }
}
